src/control/lib/control/pool.go

```diff
@@ -1057,89 +1063,212 @@ func processSCMSpaceStats(log debugLogger, filterRank filterRankFn, scmNamespace
+// The rationale for deriving space usage from ramdisk usage is that this is most likely to be the
+// limiting factor as opposed to SSD usage.
+// - Calculate mem_size as percentage of ramdisk/tmpfs free capacity.
+// - Calculate meta_size as mem_size/mem-ratio.
+// - When calculating data_size in pool…
+//   - Subtract meta_size from free blobstore space if an SSD shares META+DATA roles
+//   - Subtract WAL size from free blobstore space if an SSD shares WAL+DATA roles
+//   - Take percentage of remainder
+func getMaxPoolSizeMdOnSsd(ctx context.Context, rpcClient UnaryInvoker, createReq *PoolCreateReq, scanResp *StorageScanResp) (uint64, uint64, error) {
+	if createReq.MemRatio < 0 {
+		return 0, 0, errors.New("invalid mem-ratio, should be greater than zero")
 	}
+	if createReq.MemRatio > 1 {
+		return 0, 0, errors.New("invalid mem-ratio, should not be greater than one")
+	}
+	ranks := ranklist.RankList(createReq.Ranks)
 
-	resp, err := StorageScan(ctx, rpcClient, &StorageScanReq{Usage: true})
-	if err != nil {
-		return 0, 0, err
+	// Request parameters used to query and calculate space requirements for pool.
+	spaceReq := processSpaceReq{
+		log:               rpcClient,
+		filterRank:        newFilterRankFunc(ranks),
+		rankNVMeFreeSpace: make(rankFreeSpaceMap),
 	}
+	ramAvailBytes := uint64(math.MaxUint64)
 
-	if len(resp.HostStorage) == 0 {
-		return 0, 0, errors.New("Empty host storage response from StorageScan")
+	// Retrieve lowest RAM-disk available capacity across all pool ranks.
+	for _, key := range scanResp.HostStorage.Keys() {
+		hostStorage := scanResp.HostStorage[key].HostStorage
+
+		if hostStorage.ScmNamespaces.Usable() == 0 {
+			return 0, 0, errors.Errorf("host without ram storage: hostname=%s",
+				scanResp.HostStorage[key].HostSet.String())
+		}
+
+		usable, err := processSCMSpaceStats(spaceReq, hostStorage.ScmNamespaces)
+		if err != nil {
+			return 0, 0, err
+		}
+		rpcClient.Debugf("ram-disk: %+v, usable: %s", hostStorage.ScmNamespaces,
+			humanize.Bytes(usable))
+
+		if ramAvailBytes > usable {
+			ramAvailBytes = usable
+		}
+	}
+
+	if ramAvailBytes == math.MaxUint64 {
+		return 0, 0, errors.Errorf("no usable ram space found with rank list %s", ranks)
 	}
 
-	// Generate function to verify a rank is in the provided rank slice.
-	filterRank := newFilterRankFunc(ranks)
-	rankNVMeFreeSpace := make(rankFreeSpaceMap)
+	metaDataRolesSharedSet := false
+	metaDataRolesShared := false
+
+	for _, key := range scanResp.HostStorage.Keys() {
+		hostStorage := scanResp.HostStorage[key].HostStorage
+
+		isShared := areMetaDataRolesShared(hostStorage.NvmeDevices)
+		if metaDataRolesSharedSet && metaDataRolesShared != isShared {
+			return 0, 0, errors.Errorf("md-on-ssd role configs differ between hosts")
+		}
+		metaDataRolesShared = isShared
+		metaDataRolesSharedSet = true
+
+		if err := processNVMeSpaceStats(spaceReq, hostStorage.NvmeDevices); err != nil {
```

Comment:
naming suggestion - modifyNVMeFreeSpace

Reply:
I'm inclined to change this to return a map based on an input request to separate concerns and improve readability

Reply:
I would be in favor of that.